src/shapes/sdfgrid.cpp

```diff
@@ -562,70 +961,95 @@ class SDFGrid final : public SDF<Float, Spectrum> {
+                        (x + 0) * shape_rcp[2], (y + 1) * shape_rcp[1],
+                        (z + 0) * shape_rcp[0])));
                     bbox.expand(to_world.transform_affine(ScalarPoint3f(
-                        (x + 1) * shape_rcp[2], (y + 1) * shape_rcp[1], (z + 0) * shape_rcp[0])));
+                        (x + 1) * shape_rcp[2], (y + 1) * shape_rcp[1],
+                        (z + 0) * shape_rcp[0])));
                     bbox.expand(to_world.transform_affine(ScalarPoint3f(
-                        (x + 1) * shape_rcp[2], (y + 0) * shape_rcp[1], (z + 1) * shape_rcp[0])));
+                        (x + 1) * shape_rcp[2], (y + 0) * shape_rcp[1],
+                        (z + 1) * shape_rcp[0])));
                     bbox.expand(to_world.transform_affine(ScalarPoint3f(
-                        (x + 1) * shape_rcp[2], (y + 0) * shape_rcp[1], (z + 1) * shape_rcp[0])));
+                        (x + 1) * shape_rcp[2], (y + 0) * shape_rcp[1],
+                        (z + 1) * shape_rcp[0])));
                     bbox.expand(to_world.transform_affine(ScalarPoint3f(
-                        (x + 0) * shape_rcp[2], (y + 1) * shape_rcp[1], (z + 1) * shape_rcp[0])));
+                        (x + 0) * shape_rcp[2], (y + 1) * shape_rcp[1],
+                        (z + 1) * shape_rcp[0])));
                     bbox.expand(to_world.transform_affine(ScalarPoint3f(
-                        (x + 1) * shape_rcp[2], (y + 1) * shape_rcp[1], (z + 1) * shape_rcp[0])));
+                        (x + 1) * shape_rcp[2], (y + 1) * shape_rcp[1],
+                        (z + 1) * shape_rcp[0])));
 
                     size_t voxel_index =
                         (x + 0) + (y + 0) * (shape_v[0] - 1) +
                         (z + 0) * (shape_v[0] - 1) * (shape_v[1] - 1);
+
                     voxel_indices[count] = voxel_index;
-                    aabbs[count] = optix::BoundingBox3f(bbox);
+                    aabbs[count]         = BoundingBoxType(bbox);
 
                     count++;
                 }
             }
         }
 
-        jit_free(grid);
+        void *aabbs_ptr           = nullptr;
+        size_t *voxel_indices_ptr = nullptr;
+
+        if constexpr (dr::is_cuda_v<Float>) {
+            // TODO: async memcpy
+            aabbs_ptr =
+                jit_malloc(AllocType::Device, sizeof(BoundingBoxType) * count);
+            jit_memcpy(JitBackend::CUDA, aabbs_ptr, aabbs,
+                       sizeof(BoundingBoxType) * count);
 
-        //TODO: async memcpy
-        void *aabbs_ptr = jit_malloc(AllocType::Device, sizeof(optix::BoundingBox3f) * count);
-        jit_memcpy(JitBackend::CUDA, aabbs_ptr, aabbs, sizeof(optix::BoundingBox3f) * count);
+            voxel_indices_ptr = (size_t *) jit_malloc(AllocType::Device,
+                                                      sizeof(size_t) * count);
+            jit_memcpy(JitBackend::CUDA, voxel_indices_ptr, voxel_indices,
+                       sizeof(size_t) * count);
```

Comment:
I believe these can all be async operations: they will be added to the device queue of operations (CUDA stream). We shouldn't be reading/writing their values until the next kernel launch.